site/specification/ORCv1.md

```diff
@@ -369,6 +392,192 @@ message Metadata {
+statistics of specific columns are encrypted on disk. Column
+encryption provides fine-grain column level security even when many
+users have access to the file itself. The encryption is transparent to
+the user and the writer only needs to define which columns and
+encryption keys to use. When reading an ORC file, if the user has
+access to the keys, they will get the real data. If they do not have
+the keys, they will get the masked data.
+
+Each encrypted column in each file will have a random local key
+generated for it. Thus, even though all of the decryption happens
+locally in the reader, a malicious user that stores the key only
+enables access that column in that file. The local keys are encrypted
+by the Hadoop or Ranger Key Management Server (KMS). The encrypted
+local keys are stored in the file footer's StripeInformation.
+
+```
+message Encryption {
+  // all of the masks used in this file
+  repeated DataMask mask = 1;
+  // all of the keys used in this file
+  repeated EncryptionKey key = 2;
+  // The encrypted variants.
+  // Readers should prefer the first variant that the user has access to
+  // the corresponding key. If they don't have access to any of the keys,
+  // they should get the unencrypted data.
+  repeated EncryptionVariant variants = 3;
+}
+```
+
+## Data Masks
+
+The user's data is statically masked before writing the unencrypted
+variant. Because the masking was done statically when the file was
+written, the information about the masking is just informational.
+
+The three standard masks are:
+
+* nullify - all values become null
+* redact - replace characters with constants such as X or 9
+* sha256 - replace string with the SHA 256 of the value
+
+The default is nullify, but masks may be defined by the user. Masks
+are not allowed to change the type of the column, just the values.
+
+```
+message DataMask {
+  // the kind of masking, which may include third party masks
+  optional string name = 1;
+  // parameters for the mask
+  repeated string maskParameters = 2;
+  // the unencrypted column roots this mask was applied to
+  repeated uint32 columns = 3 [packed = true];
+}
+```
+
+## Encryption Keys
+
+In addition to the encrypted local keys, which are stored in the
+footer's StripeInformation, the file also needs to describe the master
+key that was used to encrypt the local keys. The master keys are
+described by name, their version, the encryption algorithm, and the kind
+of the KeyProvider that encrypted the local key.
+
+```
+message EncryptionKey {
+  optional string keyName = 1;
+  optional uint32 keyVersion = 2;
+  optional EncryptionAlgorithm algorithm = 3;
+  // Which KeyProvider encrypted the local keys.
+  enum KeyProviderKind {
+    UNKNOWN = 0;
+    HADOOP = 1;
+    AWS = 2;
+    GCP = 3;
+    AZURE = 4;
+  }
+  optional KeyProviderKind keyProvider = 4;
+}
+```
+
+When ORC is using the Hadoop or Ranger KMS, it generates a random local key
```

Comment:
Does this mean we need to include the KMS APIs for each provider (Ranger, AWS, etc.) inside the ORC library?

Reply:
Ranger uses the Hadoop KMS API, so we already have it. For the others, I'd propose that we build separate modules (one per a cloud provider) so that we don't end up in a dependency tangle.